tools/text/char-reader.go

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package text
+
+import (
+	"unicode"
+	"unicode/utf8"
+)
+
+// CharReader is a character reader utility used by the Ballerina lexer.
+type CharReader interface {
+	Reset(offset int)
+	Peek() rune
+	PeekN(k int) rune
+	Advance()
+	AdvanceN(k int)
+	Mark()
+	GetMarkedChars() string
+	IsEOF() bool
+}
+
+// charReaderImpl is the concrete implementation of CharReader.
+type charReaderImpl struct {
+	charBuffer       string
+	offset           int
+	charBufferLength int
+	lexemeStartPos   int
+}
+
+func CharReaderFromTextDocument(textDocument TextDocument) CharReader {
+	return CharReaderFromText(string(textDocument.ToCharArray()))
```

Comment:
Here can't we implement a direct string method for this. This would first convert the textDocuments string to rune array and then convert the same run array back to string